### PR TITLE
Performance evaluation pipeline

### PR DESCRIPTION
**Pull request by epureanudiana**

### CODE REVIEW

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+import matplotlib.pyplot as plt
+from tqdm import tqdm
+
+from blurring_as_a_service.metrics.metrics_utils import generate_binary_mask
+from blurring_as_a_service.metrics.total_blurred_area import TotalBlurredArea
+from blurring_as_a_service.metrics.yolo_labels_dataset import YoloLabelsDataset
+from blurring_as_a_service.utils.bias_category_mapper import BiasCategoryMapper
+
+
+class Size(Enum):
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:17]: Maybe ImageSize is more easy to understand out of context?

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    small = [0, 5000]
+    medium = [5000, 10000]
+    large = [10000, 1000000]
+
+
+class Class(Enum):
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:23]: Not a big fan of a class named Class. Can we improve the name?

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+            - 'sex'.
+            - 'age'.
+            - 'skin_color'.
+        Returns
+        -------
+        A list where each element is a category and its metrics, included the false negative rate.
+        """
+        result = []
+        for category in [
+            "grouped_category",
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:107]: I know I wrote the code and it's my fault :P But since you did a nice job adding the class for the ImageSize. 
Can you do the same for the SensitiveCategories?

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+        Returns
+        -------
+        A list where each element is a category value and its metrics, included the false negative rate.
+        """
+        available_categories = list(
+            {v[category_name] for k, v in self._statistics_per_category.items()}
+        )
+        all_categories = {
+            available_category: {
+                "true_positives": 0,
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:138]: Maybe also for this? And finding a better name, I have the feeling everything is `category` at the moment.

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+                "false_negatives": 0,
+                "false_negatives_small": 0,
+                "false_negatives_medium": 0,
+                "false_negatives_large": 0,
+            }
+            for available_category in available_categories
+        }
+        for value in self._statistics_per_category.values():
+            category = all_categories[value[category_name]]
+            category["true_positives"] += value["true_positives"]
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:151]: After the previous one became a class, this can become a for loop.

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+                self._calculate_false_negative_rate(
+                    false_negatives_large, true_positives_large
+                )
+                if false_negatives_large + true_positives_large > 0
+                else None
+            )
+
+            result.append(
+                {
+                    "category": category_name,
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:203]: It could be an idea to have this as an object in the future to make everything cleaner.

File: blurring_as_a_service/metrics/custom_metrics_calculator.py (hunk 1 above)
Comment (sdavrieux) [reply blurring_as_a_service/metrics/custom_metrics_calculator.py:17]: If you inherit object instead of enum the object is subscriptable and you can avoid the `.value` in `Size.small.value[0]` and just write `Size.small[0]`.

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+                        tagged_validation_content["GT_boxes"],
+                        tagged_validation_content["GT_labels"],
+                    )
+                ):
+                    category_stats = self._statistics_per_category[gt_label]
+                    bbox_area = gt_box[2] * gt_box[3] * 32000000
+                    category_stats["area"] = bbox_area
+                    if tagged_validation_content["TP_labels"][i]:
+                        category_stats["true_positives"] += 1
+                        if Size.small.value[0] <= bbox_area < Size.small.value[1]:
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:262]: It looks like repeated code, a function? :)

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+        all_grouped_categories = self._bias_category_mapper.get_all_grouped_categories()
+
+        self._statistics_per_category = {}
+        for d in all_grouped_categories:
+            for attr_1, attr_2, attr_3 in [
+                d["name"].split("/") + [""] * (3 - len(d["name"].split("/")))
+            ]:
+                if attr_1 == "licence_plate":
+                    self._statistics_per_category[d["id"]] = {
+                        "id": d["id"],
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:345]: Can we add in the if-else only the ones that are not in common and before or after add the ones are in common so we don't repeat them?

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+    return results
+
+
+def collect_tba_results_per_class_and_size(true_path, pred_path):
+    predicted_dataset = YoloLabelsDataset(folder_path=pred_path)
+    results = {}
+
+    # ====== PERSONS ====== #
+
+    for size in Size:
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:444]: Can these two blocks done using a for loop over Class?

File: blurring_as_a_service/metrics/metrics_utils.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
+from typing import List, Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+from PIL import Image
+from PIL.Image import Image as ImageType
+
+
+def process_image_labels(labels):
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/metrics_utils.py:9]: Docstring and typing ;)

File: blurring_as_a_service/metrics/custom_metrics_calculator.py [hunk 11]
```diff
@@ -0,0 +1,500 @@
...
+                        "false_negatives": 0,
+                        "false_negatives_small": 0,
+                        "false_negatives_medium": 0,
+                        "false_negatives_large": 0,
+                        "area": 0,
+                        "gt_boxes": [],
+                        "pred_boxes": [],
+                    }
+
+    def get_areas(self):
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/custom_metrics_calculator.py:386]: Docstring and typing is missing in some functions, can we add it?

File: blurring_as_a_service/metrics/metrics_utils.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+) -> Tuple[List[int], List[Tuple[float, float, float, float]]]:
+    """
+     Parses a labels file with the following normalized format: [x_center, y_center, width, height]
+
+    :param file_path: The path to the labels file to be parsed.
+    :return: A tuple with two lists: classes and bounding_boxes
+    """
+    with open(file_path, "r") as f:
+        lines = f.readlines()
+    f.close()
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/metrics_utils.py:46]: It's not necessary to close files when you use `with` in python. This is because the with when finishes call the `__exit__()` function of the class. And the open has as exit file close.

File: blurring_as_a_service/metrics/metrics_utils.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    )
+
+    return tba_true_mask, tba_pred_mask, uba_true_mask, uba_pred_mask
+
+
+def parse_labels(
+    file_path: str,
+) -> Tuple[List[int], List[Tuple[float, float, float, float]]]:
+    """
+     Parses a labels file with the following normalized format: [x_center, y_center, width, height]
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/metrics_utils.py:39]: Can we use the docstring in numpy format to have the same all around?

File: blurring_as_a_service/metrics/total_blurred_area.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        self.precision = None
+        self.tp = 0
+        self.fp = 0
+        self.tn = 0
+        self.fn = 0
+
+    def update_statistics_based_on_masks(self, true_mask, predicted_mask):
+        """
+        Computes statistics for a given pair of binary masks.
+        Args:
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/total_blurred_area.py:17]: Numpy docstring

File: blurring_as_a_service/metrics/total_blurred_area.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+            predicted_mask: numpy array of shape (height, width)
+
+        Returns:
+
+        """ """
+        Computes true positives and false positives for a given pair of binary masks.
+        :param true_mask: numpy array of shape (height, width)
+        :param pred_mask: numpy array of shape (height, width)
+        """
+        tp = np.sum((true_mask == 1) & (predicted_mask == 1))
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/total_blurred_area.py:28]: Can this two blocks be simplified in one?
For example:
`self.tp += np.sum((true_mask == 1) & (predicted_mask == 1))`

File: blurring_as_a_service/metrics/total_blurred_area.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        tn = np.sum((true_mask == 0) & (predicted_mask == 0))
+        fn = np.sum((true_mask == 1) & (predicted_mask == 0))
+
+        self.tp += tp
+        self.fp += fp
+        self.tn += tn
+        self.fn += fn
+
+    def get_statistics(self):
+        self.precision = (
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/total_blurred_area.py:39]: Should these not be in the update statistics function?
I think logically this should just get the statistics and not compute them.

File: blurring_as_a_service/metrics/upper_blurred_area.py [hunk 17]
```diff
@@ -0,0 +1,6 @@
+from metrics.total_blurred_area import TotalBlurredArea
+
+
+class UpperBlurredArea(TotalBlurredArea):
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/upper_blurred_area.py:4]: I like the idea of inheritance in case it adds functionality that the base class doesn't have.
What is the purpose in this case?

File: blurring_as_a_service/metrics/yolo_labels_dataset.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
+import os
+from pathlib import Path
+from typing import Dict
+
+import numpy as np
+import numpy.typing as npt
+from torch.utils.data import Dataset
+
+
+class YoloLabelsDataset(Dataset):
```
Comment (sdavrieux) [inline blurring_as_a_service/metrics/yolo_labels_dataset.py:10]: Docstrings and return types ;)

File: blurring_as_a_service/performace_evaluation_pipeline/components/validate_model.py [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        test=f"{mounted_dataset}/images/test/",
+        nc=2,
+        names=["person", "license_plate"],
+    )
+    with open(f"{yolo_yaml_path}/pano.yaml", "w") as outfile:
+        yaml.dump(data, outfile, default_flow_style=False)
+    os.system("cp Arial.ttf /root/.config/Ultralytics/Arial.ttf")  # nosec
+    val.run(
+        data=f"{yolo_yaml_path}/pano.yaml",
+        weights="weights/last-purple_boot_3l6p24vb.pt",
```
Comment (sdavrieux) [inline blurring_as_a_service/performace_evaluation_pipeline/components/validate_model.py:44]: The weights we decided to move them to the storage account.
You can check how it's done now in the inference pipeline ;)

File: blurring_as_a_service/performace_evaluation_pipeline/components/validate_model.py [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+        imgsz=2048,
+        batch_size=8,
+        # project="../../../outputs/runs/val",
```
Comment (sdavrieux) [inline blurring_as_a_service/performace_evaluation_pipeline/components/validate_model.py:47]: Shall we remove the comment?

File: blurring_as_a_service/performace_evaluation_pipeline/source/evaluate_with_coco.py [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+
+from metrics.customCocoEval import CustomCOCOeval
+
+
+def coco_evaluation(annotations_json, yolo_output_folder):
+
+    predictions_json = (
+        f"{yolo_output_folder}/exp/last-purple_boot_3l6p24vb_predictions.json"
+    )
+    anno = COCO(annotations_json)  # init annotations api
```
Comment (sdavrieux) [inline blurring_as_a_service/performace_evaluation_pipeline/source/evaluate_with_coco.py:13]: Not a big fan in general of shortening the names since the editors come with autocomplete.
So I would prefer annotations instead of anno for clarity.

File: blurring_as_a_service/performace_evaluation_pipeline/source/evaluate_with_coco.py [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+    pred = anno.loadRes(predictions_json)  # init predictions api
+    eval = CustomCOCOeval(anno, pred, "bbox")
+
+    # Opening JSON file
+    f = open(annotations_json)
```
Comment (sdavrieux) [inline blurring_as_a_service/performace_evaluation_pipeline/source/evaluate_with_coco.py:18]: With clause to automatically handle the file close?

File: blurring_as_a_service/performace_evaluation_pipeline/source/evaluate_with_cvt_metrics.py [hunk 23]
```diff
@@ -0,0 +1,30 @@
+from blurring_as_a_service.metrics.metrics_utils import (
+    generate_binary_mask,
+    parse_labels,
+)
+
+
+def process_image_labels(labels):
```
Comment (sdavrieux) [inline blurring_as_a_service/performace_evaluation_pipeline/source/evaluate_with_cvt_metrics.py:7]: Docstrings and typing :P

File: blurring_as_a_service/performace_evaluation_pipeline/source/evaluate_with_cvt_metrics.py [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+
+    true_classes, true_bboxes = parse_labels(labels["true"])
```
Comment (sdavrieux) [inline blurring_as_a_service/performace_evaluation_pipeline/source/evaluate_with_cvt_metrics.py:9]: This entire function or a really similar version seems to be repeated in metric utils.
Can you check and refactor to avoid repetition if it logically makes sens?

File: blurring_as_a_service/performace_evaluation_pipeline/source/get_data.py [hunk 25]
```diff
@@ -0,0 +1,8 @@
+from os import listdir
+
+
+def get_and_store_info(data_folder_path: str, output_file_path: str):
```
Comment (sdavrieux) [inline blurring_as_a_service/performace_evaluation_pipeline/source/get_data.py:4]: Mmm a file for this seems an overkill. It's also not super clear on first read what it does.
Can we think if it belongs somewhere else or together with other stuff and make the purpose clearer?

File: blurring_as_a_service/performace_evaluation_pipeline/submit_performance_evaluation_pipeline.py [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+
+@pipeline()
+def performance_evaluation_pipeline(
+    validation_data, annotations_json, yolo_yaml_path, yolo_validation_output
+):
+    performance_evaluation_flags = BlurringAsAServiceSettings.get_settings()[
+        "performance_evaluation_pipeline"
+    ]["flags"]
+
+    if performance_evaluation_flags & PipelineFlag.VALIDATE_MODEL:
```
Comment (sdavrieux) [inline blurring_as_a_service/performace_evaluation_pipeline/submit_performance_evaluation_pipeline.py:29]: Maybe the flag for validate the model we can skip it since it's the main purpose of the pipeline?
Otherwise it can also not be run. What was the idea behind it?

File: experiments/calculate_fnr.py [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+    coco_file_with_categories = (
+        "exp_with_tagged_validation_1mar/in:coco-format/validation-tagged.json"
+    )
+    markdown_output_path = (
+        f"{yolo_run_path}/custom_metrics_result_imgsz_640_with_tba.md"
+    )
+
+    metrics_calculator = CustomMetricsCalculator(source, coco_file_with_categories)
+    metrics_calculator.calculate_and_store_metrics(markdown_output_path)
+    # metrics_calculator._plot_area_distribution(path=".", plot_name="GT_bboxes_2")
```
Comment (sdavrieux) [inline experiments/calculate_fnr.py:18]: To avoid commenting and uncommenting you could set up a parameter as input called plot.
And then with an if check if needs to plot or not ;)

File: blurring_as_a_service/metrics/custom_metrics_calculator.py (hunk 3 above)
Comment (epureanudiana) [reply blurring_as_a_service/metrics/custom_metrics_calculator.py:107]: Not sure what you mean here. Because the Enum makes sense for binding small, large, medium to intervals, but in this case it is just a list of strings. 

This is what I made, is this what you meant?

```
@dataclass
class SensitiveCategories:
    values: List[str] = None

    def __post_init__(self):
        if self.values is None:
            self.values = ["grouped_category", "sex", "age", "skin_color", "licence_plate_origin", "licence_plate_color"]
```

File: blurring_as_a_service/metrics/custom_metrics_calculator.py (hunk 1 above)
Comment (epureanudiana) [reply blurring_as_a_service/metrics/custom_metrics_calculator.py:17]: Then I would be losing some nice functionalities from Enum, such as iterating through the attributes, which I am doing multiple times. 

I changed it to this for now: 
```

class ImageSize(Enum):
    small = [0, 5000]
    medium = [5000, 10000]
    large = [10000, 1000000]

    def __repr__(self):
        return self.value

    def __getitem__(self, index):
        return self.value[index]
```

File: blurring_as_a_service/metrics/metrics_utils.py (hunk 12 above)
Comment (epureanudiana) [reply blurring_as_a_service/metrics/metrics_utils.py:46]: Ahh, I did not know, thanks!

File: blurring_as_a_service/metrics/custom_metrics_calculator.py (hunk 1 above)
Comment (sdavrieux) [reply blurring_as_a_service/metrics/custom_metrics_calculator.py:17]: Good point :)

File: blurring_as_a_service/metrics/custom_metrics_calculator.py (hunk 3 above)
Comment (sdavrieux) [reply blurring_as_a_service/metrics/custom_metrics_calculator.py:107]: Mmm you're right. 
I don't know what I was exactly thinking but we can leave it as it is.

**Review by sdavrieux — APPROVED** [review]
Great job!
